store/table/table.go

```diff
@@ -0,0 +1,189 @@
+package table
+
+import (
+	"reflect"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+var _ Indexable = &Builder{}
+
+type Builder struct {
+	model         reflect.Type
+	prefixData    byte
+	indexKeyCodec IndexKeyCodec
+	afterSave     []AfterSaveInterceptor
+	afterDelete   []AfterDeleteInterceptor
+	cdc           codec.Codec
+}
+
+// NewTableBuilder creates a builder to setup a Table object.
+func NewTableBuilder(prefixData byte, model codec.ProtoMarshaler, idxKeyCodec IndexKeyCodec, cdc codec.Codec) *Builder {
+	if model == nil {
+		panic("Model must not be nil")
+	}
+	if idxKeyCodec == nil {
+		panic("IndexKeyCodec must not be nil")
+	}
+	tp := reflect.TypeOf(model)
+	if tp.Kind() == reflect.Ptr {
+		tp = tp.Elem()
+	}
+	return &Builder{
+		prefixData:    prefixData,
+		model:         tp,
+		indexKeyCodec: idxKeyCodec,
+		cdc:           cdc,
+	}
+}
+
+func (a Builder) IndexKeyCodec() IndexKeyCodec {
+	return a.indexKeyCodec
+}
+
+// RowGetter returns a type safe RowGetter.
+func (a Builder) RowGetter() RowGetter {
+	return NewTypeSafeRowGetter(a.prefixData, a.model, a.cdc)
+}
+
+// Build creates a new Table object.
+func (a Builder) Build() Table {
+	return Table{
+		model:       a.model,
+		prefix:      a.prefixData,
+		afterSave:   a.afterSave,
+		afterDelete: a.afterDelete,
+		cdc:         a.cdc,
+	}
+}
+
+// AddAfterSaveInterceptor can be used to register a callback function that is executed after an object is created and/or updated.
+func (a *Builder) AddAfterSaveInterceptor(interceptor AfterSaveInterceptor) {
+	a.afterSave = append(a.afterSave, interceptor)
+}
+
+// AddAfterDeleteInterceptor can be used to register a callback function that is executed after an object is deleted.
+func (a *Builder) AddAfterDeleteInterceptor(interceptor AfterDeleteInterceptor) {
+	a.afterDelete = append(a.afterDelete, interceptor)
+}
+
+// Table is the high level object for storage mapper functionality. Persistent entities are stored by an unique identifier
+// called `RowID`.
+// The Table struct does not enforce uniqueness of the `RowID` but expects this to be satisfied by the callers and conditions
+// to optimize Gas usage.
+type Table struct {
+	model       reflect.Type
+	prefix      byte
```

Comment:
If `prefix` identifies the protobuf object, then I think `byte` might be too small (especially if we foresee table upgrades/migrations). Maybe [2]byte gives us more capacity.

Reply:
@aaronc thoughts on that?

Reply:
cc @aaronc